lib/rules/no-this-in-template-only-components.js

```diff
@@ -0,0 +1,80 @@
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+
+const Rule = require('./base');
+
+function message(original, fixed) {
+  return (
+    `Usage of 'this' in path '${original}' is not allowed in a template-only component. ` +
+    `Use '${fixed}' if it is a named argument or create a component.js for this component.`
+  );
+}
+
+function hasContext(hbsFilePath) {
+  const paths = hbsFilePath.split(path.sep);
+  if (paths[0] === 'app' || paths[0] === 'addon') {
+    if (paths[1] === 'templates') {
+      if (paths[2] === 'components') {
+        // classic structure: /app/templates/components/foo.hbs => /app/components/foo.js
+        const moduleName = path.basename(hbsFilePath, '.hbs');
+        const jsFilePath = path.join(
+          paths[0],
+          'components',
+          ...paths.slice(3, -1),
+          `${moduleName}.js`
+        );
+        return fs.existsSync(jsFilePath);
+      } else {
+        // route template always has a context
+        return true;
+      }
+    } else if (paths[1] === 'components') {
+      // co-located structure: /app/components/foo.hbs => /app/components/foo.js
+      const moduleName = path.basename(hbsFilePath, '.hbs');
+      const jsFilePath = path.join(path.dirname(hbsFilePath), `${moduleName}.js`);
+      return fs.existsSync(jsFilePath);
+    }
+  }
+  // what other types of files?
+  return false;
+}
+
+function isFixable(original) {
+  if (original === 'this.elementId') {
+    // elementId usage in a template-only component can't be auto-fixed
+    return false;
+  }
+  return true;
+}
+
+module.exports = class NoThisInTemplateOnlyComponents extends Rule {
+  visitor() {
+    if (hasContext(this._filePath)) {
+      // template is allowed to use `this`
+      return null;
+    }
+
+    return {
+      PathExpression(path) {
+        if (path.this) {
+          const fixed = path.original.replace(/^this\./, '@');
+          if (this.mode === 'fix' && isFixable(path.original)) {
+            path.original = fixed;
```

Comment:
Hmm, I thought we also had to set `path.data = true`?

Reply:
Originally I was setting `path.data = true; path.this = false` but it still worked without those